### ISSUE TITLE
Minimal example of using Kokkos with multiple GPU through OpenMPI

### ISSUE THREAD

**Issue by Char-Aznable** [kokkos/kokkos#2096]
The existing apps have bells and whistles that often confuse me. It's better to have something that tells me the basic requirement. An example would be setting up MPI, doing some simple calculation on each GPU and then exchange data between GPUs.

**Comment by nmhamster** [comment]
@Char-Aznable - how is this any different from a standard program which uses MPI to exchange data? You don't really need to write any special code if you have a GPU-enabled MPI. The more important thing to get right is the MPI launch ensuring affinity etc.

**Comment by stanmoore1** [comment]
Yeah, you just pass the `view.data()` pointer into the MPI calls just the same as any other MPI program. If you don't have GPU-enabled MPI, then you need to get the data to the host CPU first.

**Comment by crtrott** [comment]
I think @Char-Aznable is just asking that we add a simple example to the tutorials, or examples in the Kokkos repo.

**Comment by pguthrey** [comment]
Our team is working on using this as well. We are still working on our apps using just CPUs but soon we will attempting GPU-to-GPU MPI.

**Comment by mhoemmen** [comment]
@pguthrey You'll first need to figure out whether your MPI implementation can deal with CUDA buffers.  If not, then you'll need to give MPI host-accessible memory.

**Comment by PhilMiller** [comment]
https://kokkos.github.io/kokkos-core-wiki/usecases/MPI-Halo-Exchange.html partially addresses this, except that it completely neglects the current recommendation that transferring view host memory is preferred over CUDA-aware MPI, and the relevant use of `create_mirror`, `DualView`, or `deep_copy`, and issues like `HostSpace` vs `HostPinnedSpace`.

That example is the only one I see in our documentation that mentions `MPI_Send`. There's likely better material in the tutorial slides and such, but that's far less discoverable.

**Comment by mhoemmen** [comment]
@PhilMiller wrote:
> ... the current recommendation that transferring view host memory is preferred over CUDA-aware MPI ....

please elaborate?

**Comment by PhilMiller** [comment]
https://kokkosteam.slack.com/archives/C5BGU5NDQ/p1663891781222259

In brief, Christian reports that packing into `HostPinnedSpace` buffers generally gets better performance

**Comment by PhilMiller** [comment]
Also, I think I had a typo in the quoted text, 'view' should be 'via'

**Comment by PhilMiller** [comment]
Revisiting that conversation makes me think that perhaps something in Trilinos should actually provide suitable type aliases for memory spaces that should be used for packing and communicating, rather than Kokkos doing so.

Kokkos documentation and examples should spell out the how and why of this, though.

**Comment by JBludau** [comment]
@PhilMiller I had a discussion with @crtrott after the discussion on slack. And the idea was to add a quick example with `MPI` and the `SharedHostPinned` as alias for `HostPinned`. The starting point would be the code that produced the plots Christian showed in the slack discussion. 
I just did not find the time yet to take care of drafting something

**Comment by PhilMiller** [comment]
@JBludau Those aliases have since gone in. Are you still expecting to implement this example, or have you moved on? If so, please de-assign yourself.
